[clang] Assert the enum FPOpts and LangOpts fit into the storage

fmayer opened the pull request:
Fix existing failure

fmayer committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.4

fmayer committed:
fmt

Created using spr 1.3.4

fmayer commented:
@zahiraam the default of the `ComplexRange`, `CX_None` does overflow the `2` storage bits. That means we were probably still using `CX_Full` (4 should overflow to `0`) even after https://github.com/llvm/llvm-project/pull/78330 (I am not sure why "make sure why that changed the default in the first place). Also it's impossible to manually choose `CX_None`.

zahiraam commented:
> @zahiraam the default of the `ComplexRange`, `CX_None` does overflow the `2` storage bits. That means we were probably still using `CX_Full` (4 should overflow to `0`) even after #78330 (I am not sure why why that changed the default in the first place). Also it's impossible to manually choose `CX_None`.

The only valid values for  the `-fcomplex-arithmetic` option are: `CX_Full`, `CX_Improved`, `CX_Promoted` and `CX_Basic`. 
Using `clang.exe -fcomplex-arithmetic=none t.c` would lead to this error:
`clang: error: unsupported argument 'none' to option '-fcomplex-arithmetic='`
So `2` bits of storage is enough for `ComplexRange`. `ComplexRange` values can only be 4 values above.
The value `CX_None` is the defaut value given to the `Range` when no `fcomplex-arithmetic` is used on the command line. https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/ToolChains/Clang.cpp#L2989 but it's not a valid option for `fcomplex-arithmetic`.

fmayer commented:
> > @zahiraam the default of the `ComplexRange`, `CX_None` does overflow the `2` storage bits. That means we were probably still using `CX_Full` (4 should overflow to `0`) even after #78330 (I am not sure why why that changed the default in the first place). Also it's impossible to manually choose `CX_None`.
> 
> The only valid values for the `-fcomplex-arithmetic` option are: `CX_Full`, `CX_Improved`, `CX_Promoted` and `CX_Basic`. Using `clang.exe -fcomplex-arithmetic=none t.c` would lead to this error: `clang: error: unsupported argument 'none' to option '-fcomplex-arithmetic='` So `2` bits of storage is enough for `ComplexRange`. `ComplexRange` values can only be 4 values above. The value `CX_None` is the defaut value given to the `Range` when no `fcomplex-arithmetic` is used on the command line. https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/ToolChains/Clang.cpp#L2989 but it's not a valid option for `fcomplex-arithmetic`.

Then why does this assertion trigger if I don't pass anything?

zahiraam commented:
> > > @zahiraam the default of the `ComplexRange`, `CX_None` does overflow the `2` storage bits. That means we were probably still using `CX_Full` (4 should overflow to `0`) even after #78330 (I am not sure why why that changed the default in the first place). Also it's impossible to manually choose `CX_None`.
> > 
> > 
> > The only valid values for the `-fcomplex-arithmetic` option are: `CX_Full`, `CX_Improved`, `CX_Promoted` and `CX_Basic`. Using `clang.exe -fcomplex-arithmetic=none t.c` would lead to this error: `clang: error: unsupported argument 'none' to option '-fcomplex-arithmetic='` So `2` bits of storage is enough for `ComplexRange`. `ComplexRange` values can only be 4 values above. The value `CX_None` is the defaut value given to the `Range` when no `fcomplex-arithmetic` is used on the command line. https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/ToolChains/Clang.cpp#L2989 but it's not a valid option for `fcomplex-arithmetic`.
> 
> Then why does this assertion trigger if I don't pass anything?

Not sure I understand your question?
Which assertion?
`clang.exe t.c` is triggering an assertion?

fmayer commented:
> Not sure I understand your question? Which assertion? `clang.exe t.c` is triggering an assertion?

I don't understand why you say

> The value `CX_None` is the defaut value given to the `Range` when no `fcomplex-arithmetic` is used on the command lin

but imply it does not need to be stored in the bitfield. If you patch this CL, and change the size for `ComplexRange` back, you will see it that it fails the assertion I added if you run the clang test suite.

zahiraam commented:
> > Not sure I understand your question? Which assertion? `clang.exe t.c` is triggering an assertion?
> 
> I don't understand why you say
> 
> > The value `CX_None` is the defaut value given to the `Range` when no `fcomplex-arithmetic` is used on the command lin
> 
> but imply it does not need to be stored in the bitfield. If you patch this CL, and change the size for `ComplexRange` back, you will see it that it fails the assertion I added if you run the clang test suite.

OK. I see the issue. Thanks.
I think this should be tilte [NFC][Clang] YOUR TITLE.
If you don't want to make it NFC then you need to add a LIT test :-) Not sure you can make it fail with a LIT test.

fmayer commented:
> > > Not sure I understand your question? Which assertion? `clang.exe t.c` is triggering an assertion?
> > 
> > 
> > I don't understand why you say
> > > The value `CX_None` is the defaut value given to the `Range` when no `fcomplex-arithmetic` is used on the command lin
> > 
> > 
> > but imply it does not need to be stored in the bitfield. If you patch this CL, and change the size for `ComplexRange` back, you will see it that it fails the assertion I added if you run the clang test suite.
> 
> OK. I see the issue. Thanks. I think this should be titled [NFC][Clang] YOUR TITLE. If you don't want to make it NFC then you need to add a LIT test :-) Not sure you can make it fail with a LIT test.

What do you mean? Of course you can't fail the assertion now that the offending code is fixed.

zahiraam commented:
> > > > Not sure I understand your question? Which assertion? `clang.exe t.c` is triggering an assertion?
> > > 
> > > 
> > > I don't understand why you say
> > > > The value `CX_None` is the defaut value given to the `Range` when no `fcomplex-arithmetic` is used on the command lin
> > > 
> > > 
> > > but imply it does not need to be stored in the bitfield. If you patch this CL, and change the size for `ComplexRange` back, you will see it that it fails the assertion I added if you run the clang test suite.
> > 
> > 
> > OK. I see the issue. Thanks. I think this should be titled [NFC][Clang] YOUR TITLE. If you don't want to make it NFC then you need to add a LIT test :-) Not sure you can make it fail with a LIT test.
> 
> What do you mean? Of course you can't fail the assertion now that the offending code is fixed.

I added your patch to my code base  (added only the assert, didn't change the size of storage for `ComplexRange`) and was able to confirm that the assertions go off. So you are right.
I am agreeing to merge this patch. 

However the rule is that every patch needs at least one LIT test that illustrates the issue and prove that your patch fixes it. You don't have a LIT test in your patch. So, you can either make this patch `NFC` by adding it in your title or you need to produce a LIT test.

fmayer commented:
> > > > > Not sure I understand your question? Which assertion? `clang.exe t.c` is triggering an assertion?
> > > > 
> > > > 
> > > > I don't understand why you say
> > > > > The value `CX_None` is the defaut value given to the `Range` when no `fcomplex-arithmetic` is used on the command lin
> > > > 
> > > > 
> > > > but imply it does not need to be stored in the bitfield. If you patch this CL, and change the size for `ComplexRange` back, you will see it that it fails the assertion I added if you run the clang test suite.
> > > 
> > > 
> > > OK. I see the issue. Thanks. I think this should be titled [NFC][Clang] YOUR TITLE. If you don't want to make it NFC then you need to add a LIT test :-) Not sure you can make it fail with a LIT test.
> > 
> > 
> > What do you mean? Of course you can't fail the assertion now that the offending code is fixed.
> 
> I added your patch to my code base (added only the assert, didn't change the size of storage for `ComplexRange`) and was able to confirm that the assertions go off. So you are right. I am agreeing to merge this patch.
> 
> However the rule is that every patch needs at least one LIT test that illustrates the issue and prove that your patch fixes it. You don't have a LIT test in your patch. So, you can either make this patch `NFC` by adding it in your title or you need to produce a LIT test.

Just adding the `assert` makes a whooping 20594 out of 21977 tests fail. Does it really need to be 20595? If you want to test the functionality without the assert, be my guest, but I don't think that's a useful use of time.

This patch is not NFC because it fixes buggy behaviour. If you look closely, your None value actually overflowed into unrelated bits for the FpOpts, causing who knows what behavior to change.

jansvoboda11 commented:
The idea here makes sense to me and I don't mind if it lands without "NFC" in the title, but I think enforcing this at compile time would be more robust. Could we use `static_assert` with something like `llvm::BitWidth` and supporting infrastructure from `llvm/include/ADT/BitmaskEnum.h`?

fmayer commented:
> The idea here makes sense to me and I don't mind if it lands without "NFC" in the title, but I think enforcing this at compile time would be more robust. Could we use `static_assert` with something like `llvm::BitWidth` and supporting infrastructure from `llvm/include/ADT/BitmaskEnum.h`?

That would involve annotating all the enums with the largest value, and that's a bigger change, and also prone to error, so even if we did this the `assert` here is a good idea. We can do this later.

jansvoboda11 reviewed (approved):
Ok, makes sense to fix the immediate problem now and follow up with the more robust fix later. Would be good if you could leave a FIXME on one of the `assert` calls in `LangOptions.h` outlining the potential improvement via `static_assert`. LGTM

fmayer committed:
todo

Created using spr 1.3.4
